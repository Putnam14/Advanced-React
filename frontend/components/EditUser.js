import React from 'react';
import { Query, Mutation } from 'react-apollo';
import Form from './styles/Form';
import { CURRENT_USER_QUERY, UPDATE_USER_MUTATION } from '../queries/queries';
import Error from './ErrorMessage';

class EditUser extends React.Component {
  state = {
    changes: {},
  };

  handleChange = e => {
    const { name, value } = e.target;
    const changes = {
      ...this.state.changes,
      [name]: value,
    };
    this.setState({ changes });
  };

  handleSubmit = async (e, updateUser) => {
    e.preventDefault();
    // only submit if there are real changes
    if (Object.keys(this.state.changes).length === 0) return;
    await updateUser();
    this.setState({ changes: {} });
  };

  render() {
    return (
      <Query query={CURRENT_USER_QUERY}>
        {({ data: { me }, loading }) => {
          if (loading) return <p>Loading...</p>;
          return (
            <Mutation
              mutation={UPDATE_USER_MUTATION}
              refetchQueries={[{ query: CURRENT_USER_QUERY }]}
              variables={this.state.changes}
            >
              {(updateUser, { loading, error, called, data }) => (
                <Form onSubmit={e => this.handleSubmit(e, updateUser)}>
                  <Error error={error} />
                  <fieldset disabled={loading} aria-busy={loading}>
                    {called && !error && <p data-test="updated">Updated!</p>}
                    <label htmlFor="name">
                      Name:
                      <input
                        type="text"
                        name="name"
                        defaultValue={me.name}
                        onChange={this.handleChange}
                      />
                    </label>
                    <button type="submit">Update</button>
                    <strong>me:</strong>
                    <pre>{JSON.stringify(me.name)}</pre>
                    <strong>Change:</strong>
                    <pre data-test="change">{JSON.stringify(this.state.changes)}</pre>
                  </fieldset>
                </Form>
              )}
            </Mutation>
          );
        }}
      </Query>
    );
  }
}

export default EditUser;
