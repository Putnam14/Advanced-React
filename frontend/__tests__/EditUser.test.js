import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import EditUser from '../components/EditUser';
import { mountWithApollo } from './mockMang';
import wait from 'waait';

const currentUser = {
  me: {
    name: 'Wes Bos',
  },
  refetch() {},
};

describe('<EditUser/>', () => {
  it('renders', async () => {
    const { wrapper, component } = mountWithApollo(<EditUser />);
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('Form'))).toMatchSnapshot();
  });

  it('asks you to log in when you arent logged in', () => {
    const wrapper = shallow(<EditUser />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('displays changes', async () => {
    const { wrapper } = mountWithApollo(<EditUser />);
    await wait();
    wrapper.update();
    const nameInput = wrapper.find('[name="name"]');
    nameInput.simulate('change', { target: { name: 'name', value: 'Scott' } });
    const diff = wrapper.find('pre[data-test="change"]');
    expect(diff.text()).toBe('{"name":"Scott"}');
  });

  it('calls update user when form is submitted', async () => {
    const { wrapper } = mountWithApollo(<EditUser />);
    await wait();
    wrapper.update();
    const mutation = wrapper.find('Mutation').instance();
    mutation.client.mutate = jest.fn();
    const nameInput = wrapper.find('[name="name"]');
    nameInput.simulate('change', { target: { name: 'name', value: 'Scott' } });
    wrapper.find('form').simulate('submit', { preventDefault() {} });
    expect(mutation.client.mutate).toHaveBeenCalledWithVariables({ name: 'Scott' });
  });

  it('refetches the current user after an update', async () => {
    const { wrapper } = mountWithApollo(<EditUser />);
    await wait();
    wrapper.update();
    const query = wrapper.find('Query').instance();
    const mutation = wrapper.find('Mutation').instance();
    console.log(mutation.props.mutation);
    expect('wes').toBe('finished with this one');
    // wrapper.find('form').simulate('submit', { preventDefault() {} });
    // expect(query.client.reFetchObservableQueries).toHaveBeenCalled();
  });
});
