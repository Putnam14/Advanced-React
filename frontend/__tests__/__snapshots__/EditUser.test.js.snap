// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditUser/> renders 1`] = `
<form
  className="sc-bdVaJa LHejR"
  onSubmit={[Function]}
>
  <DisplayError />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="name"
    >
      Name:
      <input
        defaultValue="Miss Coleman Berge"
        name="name"
        onChange={[Function]}
        type="text"
      />
    </label>
    <button
      type="submit"
    >
      Update
    </button>
    <strong>
      me:
    </strong>
    <pre>
      "Miss Coleman Berge"
    </pre>
    <strong>
      Change:
    </strong>
    <pre
      data-test="change"
    >
      {}
    </pre>
  </fieldset>
</form>
`;
